### PR TITLE
Add support for FluidStack-dependant colouring to Forge bucket

### LINKED ISSUE #4948: Dynamic bucket does not use FluidStack-sensitive fluid.getColor().
This prevents the enabling of the Forge Universal Bucket for fluids which rely on said sensitivity, such as fluids with NBT data that determines color.

### PR DESCRIPTION
**Pull request by bs2609**
Should fix #4948 in a way that doesn't break existing implementations. See the [comment](https://github.com/MinecraftForge/MinecraftForge/issues/4948#issuecomment-392252825) there for a description of the problem.

Firstly, the Forge bucket model is updated to provide a tint index to all quads, via updated `ItemTextureQuadConverter` methods.

Secondly, the bucket item now has an `IItemColor` implementation registered for it, which can apply colouring based on the `FluidStack` contents. The implementation used is also made available for other mods' fluid container items.

To do this in a way which doesn't cause issues for existing fluid/fluid container implementations, the model is required to opt-in to this additional colouring.

### CODE REVIEW

**Comment by asiekierka** [comment]
I slightly dislike this solution. Specifically, because we now have two ways of colouring a Fluid: the IFluidColor, and the Fluid. I feel they're partially redundant.

I believe the better solution is to put the opt-in on the bucket model itself, which is probably a lot less confusing than putting it on the fluid. Alternately, we could move to IFluidColor entirely and deprecate the Fluid's existing getColor() mechanisms, but is this really a way to go?

**Comment by asiekierka** [comment]
Looks good to me!

File: src/main/java/net/minecraftforge/client/model/ModelDynBucket.java [hunk 1]
```diff
@@ -150,16 +160,16 @@ public IBakedModel bake(IModelState state, VertexFormat format,
         {
             TextureAtlasSprite liquid = bakedTextureGetter.apply(liquidLocation);
             // build liquid layer (inside)
-            builder.addAll(ItemTextureQuadConverter.convertTexture(format, transform, liquid, fluidSprite, NORTH_Z_FLUID, EnumFacing.NORTH, fluid.getColor()));
-            builder.addAll(ItemTextureQuadConverter.convertTexture(format, transform, liquid, fluidSprite, SOUTH_Z_FLUID, EnumFacing.SOUTH, fluid.getColor()));
+            builder.addAll(ItemTextureQuadConverter.convertTexture(format, transform, liquid, fluidSprite, NORTH_Z_FLUID, EnumFacing.NORTH, tint ? fluid.getColor() : ~0, 1));
```
Comment (mezz) [inline src/main/java/net/minecraftforge/client/model/ModelDynBucket.java]: I think I'd prefer  `0xffffffff` over `~0`.

**Review by mezz — CHANGES_REQUESTED** [review]

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
The code looks good to me. Can you post a before and after picture?

**Comment by bs2609** [comment]
Here's an example using a `Fluid` implementation with these overrides:
```java
@Override
public int getColor()
{
    return Color.MAGENTA.getRGB();
}

@Override
public int getColor(FluidStack stack)
{
    return Color.CYAN.getRGB();
}
```



The universal bucket has the "new" behaviour and the Forestry can behaves the same as the existing bucket would.

**Comment by mezz** [comment]
You made this opt-in, what kind of issues do you think might come up if it was active by default?

**Comment by bs2609** [comment]
From my comment on #4948:
>This is because you want to use an IItemColor handler to have the appropriate context. Adding this to the general bucket requires removing the colour tint from the model, otherwise it'll be applied twice. However removing the tint from the model requires the other items using the model to have their own IItemColor handlers, or else the colour tint won't be applied at all.

**Comment by asiekierka** [comment]
To tl;dr: Some other mods use ModelDynBucket on their own items and expect the tint to be applied in-model, not with an IItemColor.

It's opt-in as to not break them, and should stop being opt-in in 1.13.

**Comment by jabelar** [comment]
I guess I missed it: what is the actual method for opting in to change the behavior? And the opt-in is applied to the bucket, but that setting is per-mod? Or per fluid? Like one mod can opt in and another one not opt in and both will work?

**Comment by asiekierka** [comment]
>what is the actual method for opting in to change the behavior? 

After the changes? There's a new boolean in the ModelDynBucket.

> And the opt-in is applied to the bucket, but that setting is per-mod? Or per fluid?

Per-bucket-model-kind.

**Comment by KnightMiner** [comment]
Specifically, it seems the opt in is in [this file](https://github.com/bs2609/MinecraftForge/blob/ba8f7245005677abd173b953e6d66d01f3eae548/src/main/resources/assets/forge/blockstates/dynbucket.json). This allows mods such as [my mod](https://github.com/KnightMiner/Ceramics/blob/1.12/src/main/resources/assets/ceramics/blockstates/clay_bucket.json) which use the bucket model to still work without needing updates (they will not have the NBT sensitive tints, but will still have what they had before this PR).

When I have time this evening I will verify my mod that uses the bucket model still works with this patch.

**Comment by asiekierka** [comment]
@KnightMiner Have you verified it?

**Comment by KnightMiner** [comment]
Sorry, completely forgot about this.

image

I just tested and everything works as expected. As seen in the image above on the right, both the Forge bucket and my custom bucket tint liquids property (one of the Tinkers Construct molten fluids).

For my test, I also made the test mod colored fluid return green in `getFluid(FluidStack)` in addition to setting the color to pink. As was expected, my bucket returned the default pink from `getColor()` while the Forge one used the stack sensitive green color.

**Comment by mezz** [comment]
Please run `gradle updateLicenses` to fix the license headers on your PRs that add new files.

**Pull request merged**